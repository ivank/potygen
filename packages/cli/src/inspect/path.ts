import { Tag, isNode, SqlTag } from '@potygen/ast';
import { PathItem, Path } from '../types';

const isItem =
  <T extends Tag>(predicate: (tag: SqlTag) => tag is T) =>
  (item: PathItem<Tag>): item is PathItem<T> =>
    predicate(item.tag);

export const closestParent =
  <T extends Tag>(predicate: (tag: SqlTag) => tag is T) =>
  (path: Path): T | undefined =>
    path.find(isItem(predicate))?.tag;

export const closestParentPath =
  <T extends Tag>(predicate: (tag: SqlTag) => tag is T) =>
  (path: Path): PathItem<T> | undefined =>
    path.find(isItem(predicate));

export const toPath = (tag: Tag, offset: number, path: Path = []): Path | undefined => {
  if (isNode(tag)) {
    const index = (tag.values as Tag[]).findIndex((item) => offset >= item.start && offset <= item.end + 1);

    if (index !== -1) {
      return toPath(tag.values[index], offset, [{ tag, index }, ...path]);
    } else {
      return [{ tag }, ...path];
    }
  }
  return [{ tag }, ...path];
};
