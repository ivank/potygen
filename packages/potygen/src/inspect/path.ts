/**
 * path.ts
 *
 * The ast "path" is a data structure to hold both an sql tag, as well as a "stack" of all of its parents,
 * so we can quickly understand where a tag comes from.
 */

import { findLastIndex } from '../util';
import { isNode } from '../grammar.guards';
import { Tag, SqlTag } from '../grammar.types';
import { PathItem, Path } from '../load.types';

const isItem =
  <T extends Tag>(predicate: (tag: SqlTag) => tag is T) =>
  (item: PathItem<Tag>): item is PathItem<T> =>
    predicate(item.tag);

/**
 * Find the closest parent {@link Tag} (using predicate)
 */
export const closestParent =
  <T extends Tag>(predicate: (tag: SqlTag) => tag is T) =>
  (path: Path): T | undefined =>
    path.find(isItem(predicate))?.tag;

/**
 * Find the closest parent {@link Path} (using predicate)
 */
export const closestParentPath =
  <T extends Tag>(predicate: (tag: SqlTag) => tag is T) =>
  (path: Path): PathItem<T> | undefined =>
    path.find(isItem(predicate));

/**
 * Search for the deepest tag within a tree at a given position
 */
export const toPath = (tag: Tag, offset: number, path: Path = []): Path | undefined => {
  if (isNode(tag)) {
    const index = findLastIndex<Tag>((item) => offset >= item.start && offset <= item.end + 1, tag.values);

    if (index !== -1) {
      return toPath(tag.values[index], offset, [{ tag, index }, ...path]);
    } else {
      return [{ tag }, ...path];
    }
  }
  return [{ tag }, ...path];
};
