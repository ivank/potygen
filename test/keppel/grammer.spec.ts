import { readFileSync } from 'fs';
import { join } from 'path';
import { Parser } from '@ikerin/rd-parse';
import { Keppel } from './keppel';

const keppelParser = Parser(Keppel);

describe('Keppel', () => {
  test('Keppel grammar parser generation 1', () => {
    const keppel = readFileSync(join(__dirname, '/test1.keppel'), 'utf8');
    expect(keppelParser(keppel)).toMatchSnapshot();
  });

  test('Keppel grammar parser generation 2', () => {
    const keppel = readFileSync(join(__dirname, '/test2.keppel'), 'utf8');
    expect(keppelParser(keppel)).toMatchSnapshot();
  });

  test('Keppel grammar parser error', () => {
    const keppel = readFileSync(join(__dirname, '/wrong.keppel'), 'utf8');

    expect(() => keppelParser(keppel)).toThrowErrorMatchingSnapshot();
  });
});
