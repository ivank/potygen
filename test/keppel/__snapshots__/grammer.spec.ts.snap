// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Keppel Keppel grammar parser error 1`] = `
"Unexpected token at 7:26. Remainder: '
      ]
    ]
  ]
]
// I am a comment in the end of input
"
`;

exports[`Keppel Keppel grammar parser generation 1 1`] = `
Array [
  Object {
    "body": Array [
      Object {
        "body": Array [
          Object {
            "body": Array [
              Object {
                "attributes": Array [
                  Object {
                    "name": "href",
                    "value": "#",
                  },
                  Object {
                    "name": "test",
                    "value": "My test",
                  },
                ],
                "body": Array [
                  Object {
                    "type": "free text",
                    "value": "Awesome",
                  },
                ],
                "classes": Array [
                  "brand",
                ],
                "name": "a",
                "type": "tag",
              },
              Object {
                "body": Array [
                  Object {
                    "body": Array [
                      Object {
                        "attributes": Array [
                          Object {
                            "name": "href",
                            "value": "#",
                          },
                        ],
                        "body": Array [
                          Object {
                            "type": "free text",
                            "value": "Item 1",
                          },
                        ],
                        "name": "a",
                        "type": "tag",
                      },
                    ],
                    "name": "li",
                    "type": "tag",
                  },
                  Object {
                    "body": Array [
                      Object {
                        "attributes": Array [
                          Object {
                            "name": "href",
                            "value": "#",
                          },
                        ],
                        "body": Array [
                          Object {
                            "type": "free text",
                            "value": "Item 2",
                          },
                        ],
                        "name": "a",
                        "type": "tag",
                      },
                    ],
                    "name": "li",
                    "type": "tag",
                  },
                ],
                "classes": Array [
                  "nav",
                ],
                "name": "ul",
                "type": "tag",
              },
            ],
            "classes": Array [
              "navbar-inner",
            ],
            "name": "div",
            "type": "tag",
          },
        ],
        "classes": Array [
          "navbar",
          "navbar-inverse",
          "navbar-fixed-top",
        ],
        "id": "navbar",
        "name": "div",
        "type": "tag",
      },
    ],
    "name": "body",
    "type": "tag",
  },
]
`;

exports[`Keppel Keppel grammar parser generation 2 1`] = `
Array [
  Object {
    "type": "free text",
    "value": "<!DOCTYPE html>",
  },
  Object {
    "attributes": Array [
      Object {
        "name": "lang",
        "value": "en",
      },
    ],
    "body": Array [
      Object {
        "body": Array [
          Object {
            "attributes": Array [
              Object {
                "name": "charset",
                "value": "utf-8",
              },
            ],
            "name": "meta",
            "type": "tag",
          },
          Object {
            "body": Array [
              Object {
                "type": "free text",
                "value": "Awesome nav bar example",
              },
            ],
            "name": "title",
            "type": "tag",
          },
          Object {
            "attributes": Array [
              Object {
                "name": "rel",
                "value": "stylesheet",
              },
              Object {
                "name": "href",
                "value": "http://netdna.bootstrapcdn.com/twitter-bootstrap/2.3.1/css/bootstrap-combined.no-icons.min.css",
              },
            ],
            "name": "link",
            "type": "tag",
          },
          Object {
            "attributes": Array [
              Object {
                "name": "rel",
                "value": "stylesheet",
              },
              Object {
                "name": "href",
                "value": "http://netdna.bootstrapcdn.com/font-awesome/3.1.1/css/font-awesome.min.css",
              },
            ],
            "name": "link",
            "type": "tag",
          },
        ],
        "name": "head",
        "type": "tag",
      },
      Object {
        "body": Array [
          Object {
            "body": Array [
              Object {
                "body": Array [
                  Object {
                    "attributes": Array [
                      Object {
                        "name": "href",
                        "value": "#",
                      },
                    ],
                    "body": Array [
                      Object {
                        "type": "free text",
                        "value": "Awesome",
                      },
                    ],
                    "classes": Array [
                      "brand",
                    ],
                    "name": "a",
                    "type": "tag",
                  },
                  Object {
                    "body": Array [
                      Object {
                        "body": Array [
                          Object {
                            "attributes": Array [
                              Object {
                                "name": "href",
                                "value": "#",
                              },
                            ],
                            "body": Array [
                              Object {
                                "type": "free text",
                                "value": "Item 1",
                              },
                            ],
                            "name": "a",
                            "type": "tag",
                          },
                        ],
                        "name": "li",
                        "type": "tag",
                      },
                      Object {
                        "body": Array [
                          Object {
                            "attributes": Array [
                              Object {
                                "name": "href",
                                "value": "#",
                              },
                            ],
                            "body": Array [
                              Object {
                                "type": "free text",
                                "value": "Item 2",
                              },
                            ],
                            "name": "a",
                            "type": "tag",
                          },
                        ],
                        "name": "li",
                        "type": "tag",
                      },
                    ],
                    "classes": Array [
                      "nav",
                    ],
                    "name": "ul",
                    "type": "tag",
                  },
                  Object {
                    "body": Array [
                      Object {
                        "body": Array [
                          Object {
                            "attributes": Array [
                              Object {
                                "name": "href",
                                "value": "#",
                              },
                            ],
                            "body": Array [
                              Object {
                                "body": Array [
                                  Object {
                                    "type": "free text",
                                    "value": " Logged in as: Mr. Right ",
                                  },
                                ],
                                "classes": Array [
                                  "icon-user",
                                  "icon-white",
                                ],
                                "name": "i",
                                "type": "tag",
                              },
                              Object {
                                "classes": Array [
                                  "caret",
                                ],
                                "name": "b",
                                "type": "tag",
                              },
                            ],
                            "classes": Array [
                              "dropdown-toggle",
                            ],
                            "name": "a",
                            "type": "tag",
                          },
                          Object {
                            "body": Array [
                              Object {
                                "body": Array [
                                  Object {
                                    "attributes": Array [
                                      Object {
                                        "name": "href",
                                        "value": "#",
                                      },
                                    ],
                                    "body": Array [
                                      Object {
                                        "type": "free text",
                                        "value": "Profile",
                                      },
                                    ],
                                    "name": "a",
                                    "type": "tag",
                                  },
                                ],
                                "name": "li",
                                "type": "tag",
                              },
                              Object {
                                "classes": Array [
                                  "divider",
                                ],
                                "name": "li",
                                "type": "tag",
                              },
                              Object {
                                "body": Array [
                                  Object {
                                    "attributes": Array [
                                      Object {
                                        "name": "href",
                                        "value": "#",
                                      },
                                    ],
                                    "body": Array [
                                      Object {
                                        "type": "free text",
                                        "value": "Sign out",
                                      },
                                    ],
                                    "name": "a",
                                    "type": "tag",
                                  },
                                ],
                                "name": "li",
                                "type": "tag",
                              },
                            ],
                            "classes": Array [
                              "dropdown-menu",
                            ],
                            "name": "ul",
                            "type": "tag",
                          },
                        ],
                        "classes": Array [
                          "dropdown",
                        ],
                        "name": "li",
                        "type": "tag",
                      },
                    ],
                    "classes": Array [
                      "nav",
                      "pull-right",
                    ],
                    "name": "ul",
                    "type": "tag",
                  },
                ],
                "classes": Array [
                  "navbar-inner",
                ],
                "name": "div",
                "type": "tag",
              },
            ],
            "classes": Array [
              "navbar",
              "navbar-inverse",
              "navbar-fixed-top",
            ],
            "id": "navbar",
            "name": "div",
            "type": "tag",
          },
        ],
        "name": "body",
        "type": "tag",
      },
    ],
    "name": "html",
    "type": "tag",
  },
]
`;
